### PR TITLE
WIP on issue #31

### PR DESCRIPTION
**Pull request by Sai9797**
Fixes #issue_number

**What was changed?**

Added a login and sign-up screens to the client side, and server-side support for authentication. 

**Why was it changed?**

User authentication is needed, because each user should have unique login. User sign-ups are tied to a user login.

**How was it changed?**

Server-side application uses GetHelp API token to authenticate user and get user bearer token. For development purposes, authentication can be bypassed. In this case, use developer@slu.edu as the user and any passoword. Authentication bypass only works when the server is running in DEBUG mode and DEV_USER and DEV_TOKEN config options are provided.

**Screenshots that show the changes (if applicable):**
The look of the UI needs help
Screen Shot 2023-11-28 at 11 39 21 AM

### CODE REVIEW

File: server/application/rest/work_shift.py [hunk 1]
```diff
@@ -125,7 +126,25 @@ def work_shifts():
...
+        )
+
+        if response.status_code == 200:
+            return response.json(), ResponseTypes.SUCCESS
+        else:
+            return jsonify({"message": "Invalid token"}), \
+       ResponseTypes.AUTHORIZATION_ERROR
+    except requests.RequestException as e:
+        return jsonify({"message": str(e)}), \
+            ResponseTypes.SYSTEM_ERROR
```
Comment (kate-holdener) [inline server/application/rest/work_shift.py:147]: @Sai9797 did you test this through postman? Did it work?

**Comment by Sai9797** [comment]
@kate-holdener no functionality wise i have written the code.

**Comment by Sai9797** [comment]
@kate-holdener can you review this code as when i tested i am getting empty error for get api. Is this the correct way to do.

**Comment by kate-holdener** [comment]
> @kate-holdener can you review this code as when i tested i am getting empty error for get api. Is this the correct way to do.

Can you elaborate on the the error you are getting? Since you are getting the error, it's probably not the correct way to do it. I suggest reaching out to Julia on our Slack project channel for a clarification on how to use that API.

**Comment by Sai9797** [comment]
> > @kate-holdener can you review this code as when i tested i am getting empty error for get api. Is this the correct way to do.
> 
> Can you elaborate on the the error you are getting? Since you are getting the error, it's probably not the correct way to do it. I suggest reaching out to Julia on our Slack project channel for a clarification on how to use that API.

So I am getting an empty array can you check if the code which is correct way to do.

**Comment by kate-holdener** [comment]
> > > @kate-holdener can you review this code as when i tested i am getting empty error for get api. Is this the correct way to do.
> > 
> > 
> > Can you elaborate on the the error you are getting? Since you are getting the error, it's probably not the correct way to do it. I suggest reaching out to Julia on our Slack project channel for a clarification on how to use that API.
> 
> So I am getting an empty array can you check if the code which is correct way to do.

What is the response code that you are getting?

**Comment by Sai9797** [comment]
> > > > @kate-holdener can you review this code as when i tested i am getting empty error for get api. Is this the correct way to do.
> > > 
> > > 
> > > Can you elaborate on the the error you are getting? Since you are getting the error, it's probably not the correct way to do it. I suggest reaching out to Julia on our Slack project channel for a clarification on how to use that API.
> > 
> > 
> > So I am getting an empty array can you check if the code which is correct way to do.
> 
> What is the response code that you are getting?

image

I am getting 200 response as shown in postman.

**Comment by kate-holdener** [comment]
> > > > > @kate-holdener can you review this code as when i tested i am getting empty error for get api. Is this the correct way to do.
> > > > 
> > > > 
> > > > Can you elaborate on the the error you are getting? Since you are getting the error, it's probably not the correct way to do it. I suggest reaching out to Julia on our Slack project channel for a clarification on how to use that API.
> > > 
> > > 
> > > So I am getting an empty array can you check if the code which is correct way to do.
> > 
> > 
> > What is the response code that you are getting?
> 
> image
> 
> I am getting 200 response as shown in postman.

That's the response from our server-side app. What response code are you getting from the GetHelp API when you call it to get the user information from the token?

**Comment by Sai9797** [comment]
> > > > > > @kate-holdener can you review this code as when i tested i am getting empty error for get api. Is this the correct way to do.
> > > > > 
> > > > > 
> > > > > Can you elaborate on the the error you are getting? Since you are getting the error, it's probably not the correct way to do it. I suggest reaching out to Julia on our Slack project channel for a clarification on how to use that API.
> > > > 
> > > > 
> > > > So I am getting an empty array can you check if the code which is correct way to do.
> > > 
> > > 
> > > What is the response code that you are getting?
> > 
> > 
> > image
> > I am getting 200 response as shown in postman.
> 
> That's the response from our server-side app. What response code are you getting from the GetHelp API when you call it to get the user information from the token?

image
 getting 404 error
image

**Comment by kate-holdener** [comment]
> > > > > > > @kate-holdener can you review this code as when i tested i am getting empty error for get api. Is this the correct way to do.
> > > > > > 
> > > > > > 
> > > > > > Can you elaborate on the the error you are getting? Since you are getting the error, it's probably not the correct way to do it. I suggest reaching out to Julia on our Slack project channel for a clarification on how to use that API.
> > > > > 
> > > > > 
> > > > > So I am getting an empty array can you check if the code which is correct way to do.
> > > > 
> > > > 
> > > > What is the response code that you are getting?
> > > 
> > > 
> > > image
> > > I am getting 200 response as shown in postman.
> > 
> > 
> > That's the response from our server-side app. What response code are you getting from the GetHelp API when you call it to get the user information from the token?
> 
> image getting 404 error

That's what I kind of expected. You should ask Julia from GetHelp about how to use this API endpoint (she's on our project Slack channel). I don't know the details on how to use this API, but she does.

**Comment by Sai9797** [comment]
> > > > > > > > @kate-holdener can you review this code as when i tested i am getting empty error for get api. Is this the correct way to do.
> > > > > > > 
> > > > > > > 
> > > > > > > Can you elaborate on the the error you are getting? Since you are getting the error, it's probably not the correct way to do it. I suggest reaching out to Julia on our Slack project channel for a clarification on how to use that API.
> > > > > > 
> > > > > > 
> > > > > > So I am getting an empty array can you check if the code which is correct way to do.
> > > > > 
> > > > > 
> > > > > What is the response code that you are getting?
> > > > 
> > > > 
> > > > image
> > > > I am getting 200 response as shown in postman.
> > > 
> > > 
> > > That's the response from our server-side app. What response code are you getting from the GetHelp API when you call it to get the user information from the token?
> > 
> > 
> > image getting 404 error
> 
> That's what I kind of expected. You should ask Julia from GetHelp about how to use this API endpoint (she's on our project Slack channel). I don't know the details on how to use this API, but she does.

@kate-holdener I am getting response correctly the user information for the GetAPI and when we test the get api for fetching shifts I am getting 200 response and an empty array because they are no shifts for that user so it is working as expected. Do I need to change anything.

**Comment by kate-holdener** [comment]
@Sai9797 I'm adding the Log-in functionality to the client side. In the process, I got the server to crash in the code that you added. See the stack trace below

```
Traceback (most recent call last):
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.8/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.8/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.8/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.8/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.8/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.8/site-packages/flask_cors/decorator.py", line 130, in wrapped_function
    resp = make_response(f(*args, **kwargs))
  File "/Users/holdenerea/repos/shelter_volunteers/server/application/rest/work_shift.py", line 83, in work_shifts
    user = get_user_from_token(request.headers)
  File "/Users/holdenerea/repos/shelter_volunteers/server/application/rest/work_shift.py", line 147, in get_user_from_token
    except requests.RequestException:
AttributeError: module 'requests' has no attribute 'RequestException'
```

File: server/application/rest/work_shift.py [hunk 2]
```diff
@@ -125,7 +133,23 @@ def work_shifts():
...
+    if not token:
+        return None, "No token provided"
+
+    try:
+        response = requests.get(
+            "https://api2-qa.gethelp.com/v1/users/current",
+            headers={"Authorization": f"Bearer {token}"}
+        )
+        response.raise_for_status()
+        return response.json(), None
```
Comment (kate-holdener) [inline server/application/rest/work_shift.py:146]: One problem I see is that what's being returned is the entire JSON object associated with the user. For example:
```
({'id': 6194, 'firstName': 'Kate', 'lastName': 'Holdener', 'dob': None, 'address1': None, 'address2': None, 'city': None, 'country': None, 'phone': '1234567890', 'receiveNewsletters': False, 'institutionIds': [], 'profileImageUrl': None, 'locationUpdateFrequency': 300, 'email': 'kate.holdener@slu.edu', 'state': None, 'zipCode': None}, None)

I think we'd want to identify the user by either email or their unique id. @AbhilashKotha @loganwyas what are your thoughts on this?
```

**Review by kate-holdener — DISMISSED** [review]

**Comment by kate-holdener** [comment]
@AbhilashKotha @loganwyas Can you try out this code? The client side is not very pretty, but works. If anyone is up for prettying up the login and signup pages, I'd appreciate the help. We could also do that in a separate issue. 

Also, the authentication is currently done directly from the client side (calling GetHelp api directly from client app). The login call requires the use of GetHelp API token, so I think we want to move this to our server side, and have our client side log in by contacting our server, which will use API token to contact GetHelp. In this version of the authentication code, you'll need to create a .env file in your client_app directory and define:
REACT_APP_GETHELP_AUTH_API_TOKEN=<API_TOKEN>
I'll share the API_TOKEN with you privately.

File: server/application/rest/work_shift.py (hunk 2 above)
Comment (Sai9797) [reply server/application/rest/work_shift.py:146]: May be we should try with id which is unique probably what are your thoughts @kate-holdener @AbhilashKotha @loganwyas

Comment (AbhilashKotha) [reply server/application/rest/work_shift.py:146]: Unique Id is a better option I guess

**Comment by Sai9797** [comment]
> @AbhilashKotha @loganwyas Can you try out this code? The client side is not very pretty, but works. If anyone is up for prettying up the login and signup pages, I'd appreciate the help. We could also do that in a separate issue.
> 
> Also, the authentication is currently done directly from the client side (calling GetHelp api directly from client app). The login call requires the use of GetHelp API token, so I think we want to move this to our server side, and have our client side log in by contacting our server, which will use API token to contact GetHelp. In this version of the authentication code, you'll need to create a .env file in your client_app directory and define: REACT_APP_GETHELP_AUTH_API_TOKEN=<API_TOKEN> I'll share the API_TOKEN with you privately.

I guess we can do by adding some jumbotron in the UI styles and align the form(either login/create) in the middle of the screen to make it look more better.

**Comment by AbhilashKotha** [comment]
@kate-holdener The functionality works, so I guess, we can pretty it up in a different issue. Here are the pending items that need addressing.

1. UI update for sign-in and sign-up process
2. The navbar should only display Sign-in and Sign-up when the user is not logged in.
3. Signout take user to login page

**Comment by kate-holdener** [comment]
> @kate-holdener The functionality works, so I guess, we can pretty it up in a different issue. Here are the pending items that need addressing.
> 
> 1. UI update for sign-in and sign-up process
> 2. The navbar should only display Sign-in and Sign-up when the user is not logged in.
> 3. Signout take user to login page

Thank you!

I also think that we need to give developers a "bypass authentication" mode for dev purposes. This would allow external contributors, who don't have GetHelp API key, to run our code in a dev environment. I think this would be good to add before we merge this PR to main.

**Comment by loganwyas** [comment]
@kate-holdener The code also works for me.

**Pull request merged**